Add docs for Document Picture-In-Picture API (#28557)
mdn/content: files/en-us/web/api/document_picture-in-picture_api/index.md

---
title: Document Picture-in-Picture API
slug: Web/API/Document_Picture-in-Picture_API
page-type: web-api-overview
status:
  - experimental
browser-compat: api.Window.documentPictureInPicture
---

{{SeeCompatTable}}{{DefaultAPISidebar("Document Picture-in-Picture API")}}{{securecontext_header}}

> The **Document Picture-in-Picture API** makes it possible to open an always-on-top window that can be populated with arbitrary HTML content — for example a video with custom controls or a set of streams showing the participants of a video conference call. It extends the earlier [Picture-in-Picture API for `<video>`](/en-US/docs/Web/API/Picture-in-Picture_API), which specifically enables an HTML {{htmlelement("video")}} element to be put into an always-on-top window.

Reviewer: It would be nice to update https://developer.mozilla.org/en-US/docs/Web/API/Picture-in-Picture_API as well to include the fact that the Document Picture-in-Picture API now exists as well. WDYT?

Author: Yup, agreed. I've added a note to the original PIP API landing page to cover this.

## Concepts and usage

> It is often useful to have a separate web app window available in addition to the main window that the app is running in. You might want to browse other windows while keeping certain app content in view, or you might want to give that content its own space while keeping the main app window freed up for displaying other content. You could handle this by just opening a regular new browser window, but this has two major issues:

Reviewer 2: Minor one for your consideration:

```suggestion
It is often helpful to have a different window available to a web app in addition to the main window in which the app is running. You might want to browse other windows while keeping specific app content in view, or you might want to give that content its own space while keeping the main app window freed up to display other content. You could handle this by just opening a regular new browser window, but this has two major issues:
```

1. You have to handle sharing of state information between the two windows.
2. The additional app window doesn't always stay on top, and can therefore get hidden by other windows.

> To solve these problems, web browsers have introduced APIs allowing apps to spawn an always-on-top window that is part of the same session. The first recognized use case was keeping video content playing in a separate window so that the user can continue to consume it while looking at other content. This is handled using the [Picture-in-Picture API for `<video>`](/en-US/docs/Web/API/Picture-in-Picture_API), whch is directly used on a {{htmlelement("video")}} element to place it into the separate window.

Reviewer 2: typo:

```suggestion
To solve these problems, web browsers have introduced APIs allowing apps to spawn an always-on-top window that is part of the same session. The first recognized use case was keeping video content playing in a separate window so that the user can continue to consume it while looking at other content. This is handled using the [Picture-in-Picture API for `<video>`](/en-US/docs/Web/API/Picture-in-Picture_API), which is directly used on a {{htmlelement("video")}} element to place it into the separate window.
```

See [Using the Document Picture-in-Picture API](/en-US/docs/Web/API/Document_Picture-in-Picture_API/Using) for a detailed usage guide.

## Interfaces

- {{domxref("DocumentPictureInPicture")}}
  - : Represents the document Picture-in-Picture window for the current document context.
- {{domxref("DocumentPictureInPictureEvent")}}
  - : Event object for the {{domxref("DocumentPictureInPicture/enter_event", "enter")}} event, which fires when the Picture-in-Picture window is opened.

## Extensions to other interfaces

- {{domxref("Window.documentPictureInPicture")}}
  - : Returns a reference to the document Picture-in-Picture window for the current document context.

## Examples

See [Document Picture-in-Picture API Example](https://mdn.github.io/dom-examples/document-picture-in-picture/) for a full working demo (see the full [source code](https://github.com/chrisdavidmills/dom-examples/tree/main/document-picture-in-picture) also).

## Specifications

> {{Specifications}}

Reviewer: Spec is missing. Is this normal?

Reviewer: ```
No specification data found for api.Window.documentPictureInPicture.
Check for problems with this page or contribute a missing spec_url to mdn/browser-compat-data. Also make sure the specification is included in w3c/browser-specs.
```

Author: Hrm, this usually means that the BCD updates haven't been merged into the data repo yet, but I've checked, and it is there: https://github.com/mdn/browser-compat-data/blob/main/api/Window.json#L1191. Plus, the compat data is showing up, and that is taken from the same JSON entry. 

I'm hoping this is just an issue around the preview pages being generated when it wasn't there, or something. I'll keep an eye on this.
